File: app/src/main/java/org/thoughtcrime/securesms/jobs/TrustedIntroductionsReceiveJob.java
package org.thoughtcrime.securesms.jobs;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import org.signal.core.util.logging.Log;
import org.thoughtcrime.securesms.database.SignalDatabase;
import org.thoughtcrime.securesms.database.TrustedIntroductionsDatabase;
import org.thoughtcrime.securesms.jobmanager.Data;
import org.thoughtcrime.securesms.jobmanager.Job;
import org.thoughtcrime.securesms.jobmanager.impl.NetworkConstraint;
import org.thoughtcrime.securesms.recipients.RecipientId;
import org.thoughtcrime.securesms.trustedIntroductions.TI_Data;
import org.thoughtcrime.securesms.trustedIntroductions.TI_Utils;
import org.thoughtcrime.securesms.util.Base64;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import static org.thoughtcrime.securesms.trustedIntroductions.TI_Utils.parseTIMessage;

public class TrustedIntroductionsReceiveJob extends BaseJob  {

  private static final String TAG = Log.tag(TrustedIntroductionsReceiveJob.class);

  // Factory Key
  public static final String KEY = "TIReceiveJob";

  private final RecipientId introducerId;
  private final long timestamp;
  private final String messageBody;
  private boolean bodyParsed;
  private final ArrayList<TI_Data> introductions;
  // counter keeping track of which TI_DATA has made it's way to the database
  // allows to only serialize introductions that have not yet been done if process get's interrupted
  private int                 inserts_succeeded = 0;

  // Serialization Keys
  private static final String KEY_INTRODUCER_ID = "introducer_id";
  private static final String KEY_TIMESTAMP = "timestamp";
  private static final String KEY_MESSAGE_BODY = "messageBody";
  private static final String KEY_BODY_PARSED = "bodyParsed";
  private static final String KEY_INTRODUCTIONS = "serialized_remaining_introduction_data";

  public TrustedIntroductionsReceiveJob(@NonNull RecipientId introducerId, @NonNull String messageBody, @NonNull long timestamp){
    this(introducerId,
         messageBody,
         false,
         timestamp,
         null,
         new Parameters.Builder()
                       .setQueue(introducerId.toQueueKey() + TI_Utils.serializeForQueue(messageBody) + timestamp) // TODO: Why does this need a seperate queue? Can I just enqueue each intro per introducer?
                       .setLifespan(TI_Utils.TI_JOB_LIFESPAN)
                       .setMaxAttempts(TI_Utils.TI_JOB_MAX_ATTEMPTS)
                       .addConstraint(NetworkConstraint.KEY)
                       .build());
  }

  private TrustedIntroductionsReceiveJob(@NonNull RecipientId introducerId, @NonNull String messageBody, @NonNull Boolean bodyParsed, @NonNull long timestamp, @Nullable ArrayList<TI_Data> tiData, @NonNull Parameters parameters){
    super(parameters);
    this.introducerId = introducerId;
    this.timestamp = timestamp;
    this.messageBody = messageBody;
    this.bodyParsed = bodyParsed;
    this.introductions = tiData != null ? tiData : new ArrayList<>();
  }

  /**
   * // TODO: Test serialization and deserialization
   * Serialize your job state so that it can be recreated in the future.
   */
  @NonNull @Override public Data serialize() {
    while (inserts_succeeded > 0){
      introductions.remove(0);
      inserts_succeeded--;
    }
    JSONArray serializedIntroductions = new JSONArray();
    for (TI_Data d: introductions){
      serializedIntroductions.put(d.serialize());
    }
    return new Data.Builder()
        .putString(KEY_INTRODUCER_ID, introducerId.serialize())
        .putString(KEY_MESSAGE_BODY, messageBody)
        .putBoolean(KEY_BODY_PARSED, bodyParsed)
        .putString(KEY_INTRODUCTIONS, serializedIntroductions.toString())
        .putLong(KEY_TIMESTAMP, timestamp)
        .build();
  }

  /**
   * Returns the key that can be used to find the relevant factory needed to create your job.
   */
  @NonNull @Override public String getFactoryKey() {
    return KEY;
  }

  /**
   * Called when your job has completely failed and will not be run again.
   */
  @Override public void onFailure() {
    Log.e(TAG, String.format(Locale.ENGLISH, "Failed to write introductions into the database originating from this message %s", messageBody));
  }


  @Override protected void onRun() throws Exception {
    if(!bodyParsed){
      List<TI_Data> tiData = parseTIMessage(messageBody, timestamp, introducerId);
      introductions.addAll(tiData);
      bodyParsed = true;
    }
    TrustedIntroductionsDatabase db = SignalDatabase.trustedIntroductions();
    for(TI_Data introduction: introductions){
      long result = db.incomingIntroduction(introduction);
      if (result == -1){
        // TODO: How to fail gracefully?
        throw new AssertionError(TAG + String.format("Introduction insertion for %s failed...", introduction.getIntroduceeName()));
      }
      inserts_succeeded++;
      // Testing
      Log.e(TAG, String.format("Inserted introduction of %s into the database", introduction.getIntroduceeName()));
    }
    Log.e(TAG, "TrustedIntroductionsReceiveJob completed!");
  }

  // TODO
  @Override protected boolean onShouldRetry(@NonNull Exception e) {
    return false;
  }

  public static final class Factory implements Job.Factory<TrustedIntroductionsReceiveJob> {

    @NonNull @Override public TrustedIntroductionsReceiveJob create(@NonNull Parameters parameters, @NonNull Data data) {
      // Deserialize introduction_data if present
      String serializedIntroductions = data.getString(KEY_INTRODUCTIONS);
      ArrayList<TI_Data> tiData = null;
      Log.e(TAG, serializedIntroductions);
      if (!serializedIntroductions.isEmpty()) {
        try{
          JSONArray arr = new JSONArray(serializedIntroductions);
          for (int i = 0; i < arr.length(); i++){
            // Casting directly since I'm the only one deserializing
            tiData.add(TI_Data.Deserializer.deserialize((String)arr.get(i)));
          }
        } catch (JSONException | NullPointerException e) {
          Log.e(TAG, "JSON deserialization of introductions failed!");
          e.printStackTrace();
        }
      }

      return new TrustedIntroductionsReceiveJob(RecipientId.from(data.getString(KEY_INTRODUCER_ID)),
                                                data.getString(KEY_MESSAGE_BODY),
                                                data.getBoolean(KEY_BODY_PARSED),
                                                data.getLong(KEY_TIMESTAMP),
                                                tiData,
                                                parameters);
    }
  }
}

File: app/src/main/java/org/thoughtcrime/securesms/trustedIntroductions/jobs/TrustedIntroductionsReceiveJob.kt
package org.thoughtcrime.securesms.trustedIntroductions.jobs

import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import org.signal.core.util.logging.Log.e
import org.signal.core.util.logging.Log.i
import org.signal.core.util.logging.Log.tag
import org.thoughtcrime.securesms.database.SignalDatabase.Companion.tiDatabase
import org.thoughtcrime.securesms.jobmanager.Job
import org.thoughtcrime.securesms.jobmanager.JsonJobData
import org.thoughtcrime.securesms.jobmanager.impl.NetworkConstraint
import org.thoughtcrime.securesms.jobs.BaseJob
import org.thoughtcrime.securesms.recipients.RecipientId
import org.thoughtcrime.securesms.trustedIntroductions.TI_Data
import org.thoughtcrime.securesms.trustedIntroductions.TI_Data.Deserializer.deserialize
import org.thoughtcrime.securesms.trustedIntroductions.TI_Utils
import org.thoughtcrime.securesms.trustedIntroductions.TI_Utils.constructIntroduceesFromTrustedIntrosString
import org.thoughtcrime.securesms.trustedIntroductions.TI_Utils.getIntroducerFromRawMessage
import org.thoughtcrime.securesms.trustedIntroductions.TI_Utils.serializeForQueue
import java.util.Locale
import java.util.Objects

class TrustedIntroductionsReceiveJob private constructor(
  private var introducerId: RecipientId?,
  private val messageBody: String,
  private var bodyParsed: Boolean,
  private val timestamp: Long,
  tiData: ArrayList<TI_Data>?,
  parameters: Parameters
) :
  BaseJob(parameters) {
  private val introductions = if (!tiData.isNullOrEmpty()) tiData else ArrayList()

  // counter keeping track of which TI_DATA has made it's way to the database
  // allows to only serialize introductions that have not yet been done if process gets interrupted
  private var insertsSucceeded = 0

  constructor(messageBody: String, timestamp: Long) : this(
    null,
    messageBody,
    false,
    timestamp,
    null,
    Parameters.Builder()
      .setQueue(serializeForQueue(messageBody) + timestamp)
      .setLifespan(TI_Utils.TI_JOB_LIFESPAN)
      .setMaxAttempts(TI_Utils.TI_JOB_MAX_ATTEMPTS)
      .addConstraint(NetworkConstraint.KEY)
      .build()
  )

  /**
   * Serialize your job state so that it can be recreated in the future.
   */
  override fun serialize(): ByteArray? {
    while (insertsSucceeded > 0) {
      introductions.removeAt(0)
      insertsSucceeded--
    }
    val serializedIntroductions = JSONArray()
    for (d in introductions) {
      serializedIntroductions.put(d.serialize())
    }
    return Objects.requireNonNull(
      JsonJobData.Builder()
        .putString(KEY_INTRODUCER_ID, if (introducerId == null) "NULL" else introducerId!!.serialize())
        .putString(KEY_MESSAGE_BODY, messageBody)
        .putBoolean(KEY_BODY_PARSED, bodyParsed)
        .putString(KEY_INTRODUCTIONS, serializedIntroductions.toString())
        .putLong(KEY_TIMESTAMP, timestamp)
        .build().serialize()
    )
  }

  /**
   * Returns the key that can be used to find the relevant factory needed to create your job.
   */
  override fun getFactoryKey(): String {
    return KEY
  }

  /**
   * Called when your job has completely failed and will not be run again.
   */
  override fun onFailure() {
    e(TAG, String.format(Locale.ENGLISH, "Failed to write introductions into the database originating from this message %s", messageBody))
  }


  @Throws(Exception::class)
  override fun onRun() {
    if (introducerId == null) {
      introducerId = getIntroducerFromRawMessage(messageBody)
    }
    if (!bodyParsed) {
      val tiData = constructIntroduceesFromTrustedIntrosString(messageBody, timestamp, introducerId!!)
      if (tiData == null) {
        e(TAG, "Introduction did not parse correctly, aborting!")
        return
      }
      introductions.addAll(tiData)
      bodyParsed = true
    }
    val db = tiDatabase
    for (introduction in introductions) {
      val result = db.incomingIntroduction(introduction)
      if (result == -1L) {
        // TODO: How to fail gracefully?
        e(TAG, String.format("Introduction insertion for %s failed...", introduction.introduceeName))
        //throw new AssertionError(TAG + String.format("Introduction insertion for %s failed...", introduction.getIntroduceeName()));
      }
      insertsSucceeded++
    }
    i(TAG, "TrustedIntroductionsReceiveJob completed!")
  }

  // TODO
  override fun onShouldRetry(e: Exception): Boolean {
    return false
  }

  class Factory : Job.Factory<TrustedIntroductionsReceiveJob?> {
    override fun create(parameters: Parameters, serializedData: ByteArray?): TrustedIntroductionsReceiveJob {
      // Deserialize introduction_data if present
      val data = JsonJobData.deserialize(serializedData)
      val serializedIntroductions = data.getString(KEY_INTRODUCTIONS)
      val tiData = ArrayList<TI_Data>()
      i(TAG, serializedIntroductions)
      if (serializedIntroductions.isNotEmpty()) {
        try {
          val arr = JSONArray(serializedIntroductions)
          for (i in 0 until arr.length()) {
            tiData.add(deserialize(JSONObject(arr.getString(i))))
          }
        } catch (e: JSONException) {
          // TODO: fail gracefully
          throw AssertionError("JSON deserialization of introductions failed!: " + e.message)
        } catch (e: NullPointerException) {
          throw AssertionError("JSON deserialization of introductions failed!: " + e.message)
        }
      }
      var introducer: RecipientId? = null
      if (data.getString(KEY_INTRODUCER_ID) != "NULL") {
        introducer = RecipientId.from(data.getString(KEY_INTRODUCER_ID))
      }
      return TrustedIntroductionsReceiveJob(
        introducer,
        data.getString(KEY_MESSAGE_BODY),
        data.getBoolean(KEY_BODY_PARSED),
        data.getLong(KEY_TIMESTAMP),
        tiData,
        parameters
      )
    }
  }

  companion object {
    private val TAG = String.format(TI_Utils.TI_LOG_TAG, tag(TrustedIntroductionsReceiveJob::class.java))

    // Factory Key
    const val KEY: String = "TIReceiveJob"

    // Serialization Keys
    private const val KEY_INTRODUCER_ID = "introducer_id"
    private const val KEY_TIMESTAMP = "timestamp"
    private const val KEY_MESSAGE_BODY = "messageBody"
    private const val KEY_BODY_PARSED = "bodyParsed"
    private const val KEY_INTRODUCTIONS = "serialized_remaining_introduction_data"
  }
}
